src/components/dialog/dialog.component.js

```diff
@@ -112,150 +100,144 @@ class Dialog extends Modal {
-        });
-      }
-
-      return child;
-    });
-  }
-
-  get modalHTML() {
-    let { height } = this.props;
-
-    if (height && !height.match(/px$/)) {
-      height = `${height}px`;
-    }
-
-    const dialogProps = {
-      tabIndex: 0,
-      style: {
-        minHeight: height
-      },
-      size: this.props.size,
-      fixedBottom: this.appliedFixedBottom,
-      stickyFormFooter: this.props.stickyFormFooter,
-      height: this.props.height,
-      theme: this.props.theme
-    };
-
-    if (this.props.ariaRole) {
-      dialogProps.role = this.props.ariaRole;
-    }
-
-    if (this.props.title) {
-      dialogProps['aria-labelledby'] = 'carbon-dialog-title';
-    }
-
-    if (this.props.subtitle) {
-      dialogProps['aria-describedby'] = 'carbon-dialog-subtitle';
-    }
-
-    return (
-      <DialogStyle
-        ref={ (dialog) => { this._dialog = dialog; } }
-        { ...dialogProps }
-        { ...tagComponent('dialog', { 'data-element': 'dialog', 'data-role': this.props['data-role'] }) }
-        onBlur={ this.onDialogBlur }
-      >
-        {this.dialogTitle}
-        <DialogContentStyle
-          ref={ (content) => { this._content = content; } }
-          height={ this.props.height }
-          fixedBottom={ this.appliedFixedBottom }
-        >
-          <DialogInnerContentStyle
-            ref={ (innerContent) => { this._innerContent = innerContent; } }
-            height={ this.props.height }
-          >
-            {this.renderChildren()}
-            {this.additionalContent()}
-          </DialogInnerContentStyle>
-        </DialogContentStyle>
-        {this.closeIcon}
-      </DialogStyle>
-    );
-  }
+ shouldHaveFixedBottom = () => {
+   if (!this._innerContent) return false;
+
+   const contentHeight = this._innerContent.offsetHeight + this._innerContent.offsetTop,
+       windowHeight = this.window.innerHeight - this._dialog.offsetTop - 1;
+
+   return contentHeight > windowHeight;
+ }
+
+ get dialogTitle() {
+   if (!this.props.title) return null;
+
+   const { showCloseIcon } = this.props;
+   let { title } = this.props;
+
+   if (typeof title === 'string') {
+     title = (
+       <Heading
+         title={ title }
+         titleId='carbon-dialog-title'
+         subheader={ this.props.subtitle }
+         subtitleId='carbon-dialog-subtitle'
+       />
+     );
+   }
+
+   return (
+     <DialogTitleStyle
+       showCloseIcon={ showCloseIcon }
+       ref={ (titleRef) => { this._title = titleRef; } }
+     >
+       {title}
+     </DialogTitleStyle>
+   );
+ }
+
+ get mainClasses() {
+   return classNames(
+     'carbon-dialog',
+     this.props.className
+   );
+ }
+
+ get closeIcon() {
+   if (this.props.showCloseIcon) {
+     return (
+       <Icon
+         className='carbon-dialog__close'
+         data-element='close'
+         onClick={ this.props.onCancel }
+         type='close'
+         tabIndex='0'
+         onKeyDown={ ev => (Events.isEnterKey(ev) ? this.props.onCancel() : null) }
```

Comment:
This should have role=button and should listen for Enter and Space. All close icons should be the same.